hail/src/main/scala/is/hail/io/fs/GoogleStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+package is.hail.io.fs
+
+import java.io.{ByteArrayInputStream, FileNotFoundException, InputStream, OutputStream}
+import java.net.URI
+import java.nio.ByteBuffer
+import java.nio.file.FileSystems
+
+import is.hail.utils._
+import com.google.auth.oauth2.ServiceAccountCredentials
+import com.google.cloud.{ReadChannel, WriteChannel}
+import com.google.cloud.storage.Storage.BlobListOption
+import com.google.cloud.storage.{Blob, BlobId, BlobInfo, Storage, StorageOptions}
+import org.apache.commons.io.FilenameUtils
+import org.apache.hadoop
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object GoogleStorageFS {
+  def containsWildcard(path: String): Boolean = {
+    var i = 0
+    while (i < path.length) {
+      var c = path(i)
+      if (c == '\\') {
+        i += 1
+        if (i < path.length)
+          i += 1
+        else
+          return false
+      } else if (c == '*' || c == '{' || c == '?')
+        return true
+
+      i += 1
+    }
+
+    return false
+  }
+
+  def getBucketPath(filename: String): (String, String) = {
+    val uri = new URI(filename)
+
+    val scheme = uri.getScheme
+    assert(scheme != null && scheme == "gs", uri.getScheme)
+
+    val bucket = uri.getHost
+    assert(bucket != null)
+
+    var path = uri.getPath
+    if (path.nonEmpty && path.head == '/')
+      path = path.drop(1)
+
+    (bucket, path)
+  }
+
+  def dropTrailingSlash(path: String): String = {
```

Comment:
This can be simplified to if(path.endWith("/")) path.substring(0, path.lastIndexOf("/")) else path

Reply:
That won't work, it only pops off the last slash.  I want last index of not slash.  The invariant for the function is that the last character is not a slash.

Reply:
Sorry, .indexOf, not last index of

`if(path.endWith("/")) path.substring(0, path.indexOf("/") - 1)`

Reply:
IndexOf returns the first occurrence, so this doesn't work if the path contains multiple components like "a/b/c/".

Reply:
gotcha